bot/exts/info/tags.py

```diff
@@ -24,145 +29,225 @@
 FOOTER_TEXT = f"To show a tag, type {constants.Bot.prefix}tags <tagname>."
 
 
+class COOLDOWN(enum.Enum):
+    """Sentinel value to signal that a tag is on cooldown."""
+
+    obj = object()
+
+
+class TagIdentifier(NamedTuple):
+    """Stores the group and name used as an identifier for a tag."""
+
+    group: Optional[str]
+    name: str
+
+    def get_fuzzy_score(self, fuzz_tag_identifier: TagIdentifier) -> float:
+        """Get fuzzy score, using `fuzz_tag_identifier` as the identifier to fuzzy match with."""
+        if self.group is None:
+            if fuzz_tag_identifier.group is None:
+                # We're only fuzzy matching the name
+                group_score = 1
+            else:
+                # Ignore tags without groups if the identifier contains a group
+                return .0
+        else:
+            if fuzz_tag_identifier.group is None:
+                # Ignore tags with groups if the identifier does not have a group
+                return .0
+            else:
+                group_score = _fuzzy_search(fuzz_tag_identifier.group, self.group)
+
+        fuzzy_score = group_score * _fuzzy_search(fuzz_tag_identifier.name, self.name) * 100
+        if fuzzy_score:
+            log.trace(f"Fuzzy score {fuzzy_score:=06.2f} for tag {self!r} with fuzz {fuzz_tag_identifier!r}")
+        return fuzzy_score
+
+    def __str__(self) -> str:
+        if self.group is not None:
+            return f"{self.group} {self.name}"
+        else:
+            return self.name
+
+    @classmethod
+    def from_string(cls, string: str) -> TagIdentifier:
+        """Create a `TagIdentifier` instance from the beginning of `string`."""
+        split_string = string.removeprefix(constants.Bot.prefix).split(" ", maxsplit=2)
+        if len(split_string) == 1:
+            return cls(None, split_string[0])
+        else:
+            return cls(split_string[0], split_string[1])
+
+
+class Tag:
+    """Provide an interface to a tag from resources with `file_content`."""
+
+    def __init__(self, content_path: Path):
+        post = frontmatter.loads(content_path.read_text("utf8"))
+        self.file_path = content_path
+        self.content = post.content
+        self.metadata = post.metadata
+        self._restricted_to: set[int] = set(self.metadata.get("restricted_to", ()))
+        self._cooldowns: dict[discord.TextChannel, float] = {}
+
+    @property
+    def embed(self) -> Embed:
+        """Create an embed for the tag."""
+        embed = Embed.from_dict(self.metadata.get("embed", {}))
+        embed.description = self.content
+        return embed
+
+    def accessible_by(self, member: discord.Member) -> bool:
+        """Check whether `member` can access the tag."""
+        return bool(
+            not self._restricted_to
+            or self._restricted_to & {role.id for role in member.roles}
```

Comment:
This is actually a bit sad when you think about it haha

Discord.py uses an internal array to store role IDs attached to a member, `.roles` is a property that gets all of the roles from cache by their ID. We then use a set comprehension to convert it to a set of.. role IDs.